janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMultiQueryStep.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Step;
+import org.apache.tinkerpop.gremlin.process.traversal.Traverser;
+import org.apache.tinkerpop.gremlin.process.traversal.Traverser.Admin;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+
+import com.google.common.collect.Lists;
+
+/**
+ * This step can be injected before a traversal parent, such as a union, and will cache the
+ * starts sent to the parent. The traversal parent will drip feed those starts into its
+ * child traversal. If the initial step of that child supports multiQuery then its faster
+ * if initialised with all the starts than just one at a time, so this step allows it to
+ * request the full set of starts from this step when initialising itself.
+ */
+public final class JanusGraphMultiQueryStep extends AbstractStep<Vertex, Vertex> {
+
+    private final Set<Traverser.Admin<Vertex>> cachedStarts = new HashSet<Traverser.Admin<Vertex>>();
+    private boolean cachedStartsAccessed = false;
+
+    public JanusGraphMultiQueryStep(Step<Vertex,?> originalStep) {
+        super(originalStep.getTraversal());
+    }
+
+    @Override
+    protected Admin<Vertex> processNextStart() throws NoSuchElementException {
+        Admin<Vertex> start = this.starts.next();
+        if (!cachedStarts.contains(start))
+        {
+            if (cachedStartsAccessed == true) {
+                cachedStarts.clear();
+                cachedStartsAccessed = false;
+            }
+            final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
+            starts.forEachRemaining(s -> { newStarters.add(s); });
+            starts.add(newStarters.iterator());
+            cachedStarts.addAll(newStarters);
```

Comment:
I just think that it is an unnecessary loop. It isn't critical but I think we can omit it like here:
```
final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
while (starts.hasNext()){
    Admin<Vertex> vertexAdmin = starts.next();
    newStarters.add(vertexAdmin);
    cachedStarts.add(vertexAdmin);
}
starts.add(newStarters.iterator());
cachedStarts.add(start);
```
Also, could you explain the line `starts.add(newStarters.iterator());`? I am not sure why do we need it. I initially sought that you want to expend the `starts` iterator with same elements. Then I checked that the iterator uses `TraverserSet` for its elements. So, when you actually call `starts.add(newStarters.iterator());` it actually executes the code like `existing.merge(traverser);` because all elements from `newStarters` already exists in `starts`. So we actually can replace the code to be like:
```
while (starts.hasNext()){
    Admin<Vertex> vertexAdmin = starts.next();
    cachedStarts.add(vertexAdmin);
    vertexAdmin.merge(vertexAdmin);
}
cachedStarts.add(start);
```
Am I missing something?

Reply:
Also, I didn't check but I assume that `merge` method just rewrites all changes into the existing object. So, in current situation it does nothing (need to check). If so, the code can be simplified like:
```
while (starts.hasNext()){
    cachedStarts.add(starts.next());
}
cachedStarts.add(start);
```

Reply:
Sorry, I missed that `starts` iterator won't be reusable after that

Reply:
I think then the next code can be used:
```
final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
starts.forEachRemaining(s -> {
    newStarters.add(s);
    cachedStarts.add(s);
});
starts.add(newStarters.iterator());
cachedStarts.add(start);
```

Reply:
I followed the patter used in [JanusGraphVertexStep. initialize()](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphVertexStep.java#L94) and it confused me at first too but as I think you've just spotted the [starts.next()](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java#L53) is actually removing the start. So after we've looked at all the starts we need to call `starts.add(newStarters.iterator());` to put them all back again. 

Have just pushed up a change for the other comments, let me know if you think we need changes here too.

Reply:
Think our comments crossed each other - your last one looks good will change it to do that.